File: backend/src/adapters/adapters.service.ts
import { Injectable } from '@nestjs/common';
import { AwsBedrockAdapter } from './aws-bedrock.adapter';
import { AzureOpenAIAdapter } from './azure-openai.adapter';
import { GoogleVertexAdapter } from './google-vertex.adapter';
import { MockAdapter } from './mock.adapter';
import { OpenAIAdapter } from './openai.adapter';
import { ProviderInvocationResult } from './types';

@Injectable()
export class AdaptersService {
  async invokeProvider(
    providerType: string,
    credentials: string,
    model: string,
    payload: Record<string, unknown>
  ): Promise<ProviderInvocationResult> {
    const normalized = (providerType || 'openai').toLowerCase();
    const parsedCredentials = this.safeParse(credentials);

    const adapter = this.resolveAdapter(normalized);
    return adapter.invoke(parsedCredentials, model, payload);
  }

  private resolveAdapter(providerType: string) {
    switch (providerType) {
      case 'azure':
      case 'azure_openai':
      case 'azure-openai':
        return new AzureOpenAIAdapter();
      case 'aws':
      case 'bedrock':
      case 'aws-bedrock':
        return new AwsBedrockAdapter();
      case 'google':
      case 'gcp':
      case 'vertex':
      case 'vertex-ai':
        return new GoogleVertexAdapter();
      case 'mock':
        return new MockAdapter();
      case 'openai':
      default:
        return new OpenAIAdapter();
    }
  }

  private safeParse(credentials: string) {
    try {
      return JSON.parse(credentials) as Record<string, unknown>;
    } catch (error) {
      throw new Error('Invalid credentials format, must be JSON');
    }
  }
}

File: backend-java/src/main/java/com/neria/manager/adapters/AdaptersService.java
package com.neria.manager.adapters;

import com.fasterxml.jackson.databind.ObjectMapper;
import java.util.Map;
import org.springframework.stereotype.Service;

@Service
public class AdaptersService {
  private final ObjectMapper objectMapper;

  public AdaptersService(ObjectMapper objectMapper) {
    this.objectMapper = objectMapper;
  }

  public ProviderInvocationResult invokeProvider(
      String providerType, String credentialsJson, String model, Map<String, Object> payload) {
    String normalized = providerType == null ? "openai" : providerType.toLowerCase();
    Map<String, Object> credentials = safeParse(credentialsJson);
    ProviderAdapter adapter = resolveAdapter(normalized);
    return adapter.invoke(credentials, model, payload);
  }

  private ProviderAdapter resolveAdapter(String providerType) {
    return switch (providerType) {
      case "azure", "azure_openai", "azure-openai" -> new AzureOpenAIAdapter(objectMapper);
      case "aws", "bedrock", "aws-bedrock" -> new AwsBedrockAdapter();
      case "google", "gcp", "vertex", "vertex-ai" -> new GoogleVertexAdapter();
      case "mock" -> new MockAdapter();
      case "openai" -> new OpenAIAdapter(objectMapper);
      default -> new OpenAIAdapter(objectMapper);
    };
  }

  private Map<String, Object> safeParse(String credentials) {
    if (credentials == null || credentials.isBlank()) {
      return Map.of();
    }
    try {
      return objectMapper.readValue(credentials, Map.class);
    } catch (Exception ex) {
      throw new IllegalArgumentException("Invalid credentials format, must be JSON");
    }
  }
}
